hack/generator/pkg/codegen/pipeline_convert_allof_and_oneof_to_objects.go

```diff
@@ -0,0 +1,542 @@
+	if leftOptional, ok := left.(*astmodel.OptionalType); ok {
+		// is this wrong? it feels wrong, but needed for {optional{enum}, string}
+		return s.intersectTypes(leftOptional.Element(), right)
+	}
+
+	return nil, nil
+}
+
+func (s synthesizer) handleResourceType(left astmodel.Type, right astmodel.Type) (astmodel.Type, error) {
+	if leftResource, ok := left.(*astmodel.ResourceType); ok {
+		if rightResource, ok := right.(*astmodel.ResourceType); ok {
+			// merge two resources: merge spec/status
+
+			spec, err := s.intersectTypes(leftResource.SpecType(), rightResource.SpecType())
+			if err != nil {
+				return nil, err
+			}
+
+			var status astmodel.Type
+			if leftResource.StatusType() != nil && rightResource.StatusType() != nil {
+				status, err = s.intersectTypes(leftResource.StatusType(), rightResource.StatusType())
+				if err != nil {
+					return nil, err
+				}
+			} else if leftResource.StatusType() != nil {
+				status = leftResource.StatusType()
+			} else {
+				status = rightResource.StatusType()
+			}
+
+			return leftResource.WithSpec(spec).WithStatus(status), nil
+		}
+
+		if s.specOrStatus == chooseStatus {
+			if leftResource.StatusType() != nil {
+				newT, err := s.intersectTypes(leftResource.StatusType(), right)
+				if err != nil {
+					return nil, err
+				}
+
+				return leftResource.WithStatus(newT), nil
+			} else {
+				return leftResource.WithStatus(right), nil
+			}
+		} else if s.specOrStatus == chooseSpec {
+			newT, err := s.intersectTypes(leftResource.SpecType(), right)
+			if err != nil {
+				return nil, err
+			}
+
+			return leftResource.WithSpec(newT), nil
+		} else {
+			panic("invalid specOrStatus")
+		}
+	}
+
+	return nil, nil
+}
+
+func (s synthesizer) handleOptionalOptional(left astmodel.Type, right astmodel.Type) (astmodel.Type, error) {
+	// if both optional merge their contents and put back in an optional
+	if leftOptional, ok := left.(*astmodel.OptionalType); ok {
+		if rightOptional, ok := right.(*astmodel.OptionalType); ok {
+			result, err := s.intersectTypes(leftOptional.Element(), rightOptional.Element())
+			if err != nil {
+				return nil, err
+			}
+
+			return astmodel.NewOptionalType(result), nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (s synthesizer) handleMapMap(left astmodel.Type, right astmodel.Type) (astmodel.Type, error) {
+	if leftMap, ok := left.(*astmodel.MapType); ok {
+		if rightMap, ok := right.(*astmodel.MapType); ok {
+			keyType, err := s.intersectTypes(leftMap.KeyType(), rightMap.KeyType())
+			if err != nil {
+				return nil, err
+			}
+
+			valueType, err := s.intersectTypes(leftMap.ValueType(), rightMap.ValueType())
+			if err != nil {
+				return nil, err
+			}
+
+			return astmodel.NewMapType(keyType, valueType), nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (s synthesizer) handleObjectObject(left astmodel.Type, right astmodel.Type) (astmodel.Type, error) {
+	if leftObj, ok := left.(*astmodel.ObjectType); ok {
```

Comment:
minor: Channeling my inner @babbageclunk you could invert some of these checks to reduce nesting.

A lot of the methods are so short it's not really worth it, but this one would benefit from it probably, as would a few of the others.